File: javascript/src/0_99/02_even_fibonacci_numbers/02_even_fibonacci_numbers.ts
/*
  Link: https://projecteuler.net/problem=2

  Problem 2: Even Fibonacci Numbers

  My solution will need:
    - a memoized fibonacci function that can calculate up to 4 million without timing out
    - to be able to summate all even numbers within the fibonacci sequence

  The steps I am thinking about taking are:
    1. Calculate an array of fibonacci numbers.
    2. call Array.prototype.reduce() to summate all even numbers.
*/

// ---------------------------------------------------------------------------

const fib = (max: number = 4000000) => {
  // This object is used to store calculated
  const cache = {}


}

File: 02_even_fibonacci_numbers.rb
def even_fibonacci_numbers
    array = []
    n = 1
    while fib(n) < 4_000_000
        array << fib(n)
        n += 1
    end
    array.select { |number| number.even? }.reduce(:+)
end

def fib_memo(n, memo)
    memo[n] ||= fib_memo(n-1, memo) + fib_memo(n-2, memo)
end

def fib(n)
    fib_memo(n, [0, 1])
end

p even_fibonacci_numbers == 4613732